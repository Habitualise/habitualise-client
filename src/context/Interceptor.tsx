import {createContext, useState, useEffect, useMemo} from 'react';
import {useAuth0} from 'react-native-auth0';
import {axios} from '../lib/axios';

export const InterceptorContext = createContext({isTokenSet: false});

const Interceptor = ({children}) => {
  const {loading, getCredentials} = useAuth0();
  const [accessToken, setAccessToken] = useState();

  useEffect(() => {
    const getAccessToken = async () => {
      try {
        const token = await getCredentials();
        if (token && token.accessToken) {
          setAccessToken(token.accessToken);
        }
      } catch (error) {
        console.log(error);
      }
    };
    if (!loading) {
      getAccessToken();
    }
  }, [loading, getCredentials]);

  useMemo(() => {
    axios.interceptors.request.use(
      async config => {
        const getAccessToken = async () => {
          try {
            const token = await getCredentials();
            if (token && token.accessToken) {
              setAccessToken(token.accessToken);
            }
          } catch (error) {
            console.log(error);
          }
        };
        if (!loading) {
          getAccessToken();
        }
        if (accessToken) {
          console.log(
            '=== setting axios interceptor accessToken Interceptor.tsx [44] ===',
            accessToken,
          );
          config.headers.Authorization = `Bearer ${accessToken}`;
        }
        return config;
      },
      error => {
        return Promise.reject(error);
      },
    );
  }, [loading, getCredentials, accessToken]);

  return (
    <InterceptorContext.Provider value={{isTokenSet: !!accessToken}}>
      {children}
    </InterceptorContext.Provider>
  );
};

export default Interceptor;
