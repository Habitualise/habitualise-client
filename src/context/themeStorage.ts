import {MMKVLoader} from 'react-native-mmkv-storage';

const themeStorage = new MMKVLoader().withInstanceID('theme').initialize();

export const setTheme = (isDark: boolean) => {
  themeStorage.setString('theme', isDark ? 'dark' : 'light');
};

export const getTheme = () => {
  const theme = themeStorage.getString('theme');
  return theme === 'dark';
};

export default themeStorage;
