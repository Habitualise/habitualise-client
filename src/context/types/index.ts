export interface Habit {
  id: string;
  name: string;
  iconName: string;
  colour: HabitColor;
  isCompleted: boolean;
  completionPercentage: number;
  active: boolean;
  daysDue: number[];
  completionHistory: boolean[]; // DONE, SKIPPED, MISSED - one for each day, last element is the most recent (today)
}

export type HabitColor =
  | 'red'
  | 'green'
  | 'blue'
  | 'yellow'
  | 'purple'
  | 'orange'
  | 'grey';

export interface UserBE {
  name: string;
}
