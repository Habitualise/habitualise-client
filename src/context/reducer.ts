import {IState} from '@/types';

export const ACTIONS = {
  SET_ALL_HABITS: 'SET_ALL_HABITS',
  TOGGLE_HABIT_IS_COMPLETED: 'TOGGLE_HABIT_IS_COMPLETED',
  HANDLE_LOGOUT: 'HANDLE_LOGOUT',
};

export const reducer = (state: IState, action) => {
  switch (action.type) {
    case ACTIONS.SET_ALL_HABITS:
      // action.payload should be an array of habits
      // TODO: add some validation here
      return {...state, habits: action.payload};
    case ACTIONS.HANDLE_LOGOUT:
      return {...state, habits: []};
    case ACTIONS.TOGGLE_HABIT_IS_COMPLETED: {
      // action.payload should be the id of the habit
      // scan through habits and find the one that matches the id
      // then toggle the isCompleted property
      const updatedHabits = state.habits.map(habit => {
        if (habit.id === action.payload) {
          const updatedIsCompleted = !habit.isCompleted;
          // today's date e.g. '2023-02-13' to add to or remove from lastSevenCompletions array
          const today = new Date();
          const todayString = today.toISOString().split('T')[0];

          if (updatedIsCompleted) {
            habit.lastSevenCompletions.push(todayString);
          } else {
            habit.lastSevenCompletions = habit.lastSevenCompletions.filter(
              date => date !== todayString,
            );
          }

          return {...habit, isCompleted: !habit.isCompleted};
        }
        return habit;
      });
      return {...state, habits: updatedHabits};
    }
    default:
      return state;
  }
};
