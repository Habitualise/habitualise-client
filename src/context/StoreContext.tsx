import * as React from 'react';
import {State} from './reducer';

export * from './reducer';

interface HabitsStore {
  state: State;
  dispatch: React.Dispatch<any>;
}

const StoreContext = React.createContext<HabitsStore>({
  state: {habits: []},
  dispatch: () => {},
});

const initState = {
  habits: [],
};

interface StoreContextProviderProps {
  children: React.ReactNode;
  reducer: React.Reducer<any, any>;
}

export const StoreContextProvider = ({
  children,
  reducer,
}: StoreContextProviderProps) => {
  const [state, dispatch] = React.useReducer(reducer, initState);
  const memoizedValue = React.useMemo(() => ({state, dispatch}), [state]);
  return (
    <StoreContext.Provider value={memoizedValue}>
      {children}
    </StoreContext.Provider>
  );
};

export const useStore = () => React.useContext(StoreContext);
