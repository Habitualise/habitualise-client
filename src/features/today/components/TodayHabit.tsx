import React from 'react';
import {Divider, Text} from 'react-native-paper';
import {StyleSheet, View} from 'react-native';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';

import {themeColors} from '@app/theme';
import {HabitIcon} from '@app/components/HabitIcon';
import {ACTIONS, useStore} from '@app/context/StoreContext';
import {formatDaysDue} from '../utils/formatDaysDue';
import {PressableCard} from '@app/components/PressableCard';
import {HabitColor} from '@app/context/types';

interface TodayHabitProps {
  id: string;
  name: string;
  iconName: string;
  colour: HabitColor;
  isCompletedToday: boolean;
  daysDue: number[];
}

export const TodayHabit = (props: TodayHabitProps) => {
  const {id, name, iconName, colour, isCompletedToday, daysDue} = props;
  const {dispatch} = useStore();

  const toggleHabitCompleted = () => {
    dispatch({
      type: ACTIONS.TOGGLE_HABIT_IS_COMPLETED,
      payload: id,
    });
  };

  return (
    <>
      <PressableCard
        onPress={toggleHabitCompleted}
        isGreyedOut={isCompletedToday}>
        <HabitIcon iconName={iconName} colour={colour} />
        <View style={{flex: 1}}>
          <Text style={styles(isCompletedToday).habitName}>{name}</Text>
          <Text
            style={styles(isCompletedToday).habitDaysDue}
            variant={'bodySmall'}>
            {formatDaysDue(daysDue)}
          </Text>
        </View>
        <MaterialCommunityIcons
          name={isCompletedToday ? 'check-circle' : 'circle-outline'}
          color={
            isCompletedToday
              ? themeColors.successGreyedOut
              : themeColors.grey[400]
          }
          size={25}
        />
      </PressableCard>
      <Divider />
    </>
  );
};

const styles = (isCompletedToday = false) =>
  StyleSheet.create({
    habitName: {
      marginBottom: 4,
      fontWeight: 'bold',
      textDecorationLine: isCompletedToday ? 'line-through' : 'none',
      color: isCompletedToday ? themeColors.grey[500] : themeColors.grey[900],
    },
    habitDaysDue: {
      color: isCompletedToday ? themeColors.grey[400] : themeColors.grey[500],
    },
  });
