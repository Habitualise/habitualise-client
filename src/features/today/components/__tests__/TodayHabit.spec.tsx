import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react-native';

import { TodayHabit } from '../TodayHabit';
import { StoreContextProvider, ACTIONS } from '@/context/StoreContext';

const toggleHabit = jest.fn();

const reducer = jest.fn((state, action) => {
  switch (action.type) {
    case ACTIONS.TOGGLE_HABIT_IS_COMPLETED:
      toggleHabit();
      return state;
    default:
      return state;
  }
});

describe('TodayHabit', () => {
  it('should match snapshot for not completed', () => {
    const { toJSON } = render(
      <StoreContextProvider reducer={reducer}>
        <TodayHabit
          colour="rgb(0, 0, 0"
          daysDue={[1, 2, 3, 4, 5, 6, 7]}
          iconName="abacus"
          id="1"
          isCompleted={false}
          name="test"
        />
      </StoreContextProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should match snapshot for completed', () => {
    const { toJSON } = render(
      <StoreContextProvider reducer={reducer}>
        <TodayHabit
          colour="rgb(0, 0, 0"
          daysDue={[1, 2, 3]}
          iconName="abacus"
          id="1"
          isCompleted={true}
          name="test"
        />
      </StoreContextProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should call dispatch when pressed', () => {
    render(
      <StoreContextProvider reducer={reducer}>
        <TodayHabit
          colour="rgb(0, 0, 0"
          daysDue={[0, 1, 2, 3, 4, 5, 6]}
          iconName="abacus"
          id="1"
          isCompleted={false}
          name="test"
        />
      </StoreContextProvider>,
    );

    fireEvent.press(screen.getByText('test'));
    expect(toggleHabit).toHaveBeenCalled();
  });
});
