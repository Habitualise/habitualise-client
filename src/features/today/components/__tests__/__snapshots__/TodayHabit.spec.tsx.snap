// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodayHabit should match snapshot for completed 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0,0,0,0)",
      },
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0.04)",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0",
          "borderRadius": 50,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginRight": 16,
          "padding": 5,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "rgb(0, 0, 0",
              "fontSize": 32,
            },
            undefined,
            {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        󱛠
      </Text>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontWeight": "400",
              "letterSpacing": 0,
            },
            {
              "writingDirection": "ltr",
            },
            {
              "color": "rgb(158, 158, 158)",
              "fontWeight": "bold",
              "marginBottom": 4,
              "textDecorationLine": "line-through",
            },
          ]
        }
      >
        test
      </Text>
      <Text
        style={
          [
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 12,
              "fontWeight": "400",
              "letterSpacing": 0.4,
              "lineHeight": 16,
            },
            {
              "textAlign": "left",
            },
            {
              "writingDirection": "ltr",
            },
            {
              "color": "rgb(189, 189, 189)",
            },
          ]
        }
      >
        Mon, Tue, Wed
      </Text>
    </View>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "rgb(121, 184, 149)",
            "fontSize": 25,
          },
          undefined,
          {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      󰗠
    </Text>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(202, 196, 208, 1)",
          "height": 0.5,
        },
        undefined,
        false,
        false,
        undefined,
      ]
    }
  />
</View>
`;

exports[`TodayHabit should match snapshot for not completed 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(0,0,0,0)",
      },
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0)",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0",
          "borderRadius": 50,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginRight": 16,
          "padding": 5,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "rgb(0, 0, 0",
              "fontSize": 32,
            },
            undefined,
            {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        󱛠
      </Text>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontWeight": "400",
              "letterSpacing": 0,
            },
            {
              "writingDirection": "ltr",
            },
            {
              "color": "rgb(33, 33, 33)",
              "fontWeight": "bold",
              "marginBottom": 4,
              "textDecorationLine": "none",
            },
          ]
        }
      >
        test
      </Text>
      <Text
        style={
          [
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 12,
              "fontWeight": "400",
              "letterSpacing": 0.4,
              "lineHeight": 16,
            },
            {
              "textAlign": "left",
            },
            {
              "writingDirection": "ltr",
            },
            {
              "color": "rgb(158, 158, 158)",
            },
          ]
        }
      >
        Every day
      </Text>
    </View>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "rgb(189, 189, 189)",
            "fontSize": 25,
          },
          undefined,
          {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      󰝦
    </Text>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(202, 196, 208, 1)",
          "height": 0.5,
        },
        undefined,
        false,
        false,
        undefined,
      ]
    }
  />
</View>
`;
